### DISCUSSION TITLE
Charts in Callout Boxes Overflow Their Boundaries in Typst Format

### DISCUSSION THREAD

**Discussion by produnis** [quarto-dev/quarto-cli#11573] [Q&A] [answered]
### Description

When generating documents in the Typst format with Quarto, charts placed inside callout boxes are rendered in a way that they overflow the boundaries of the box. This negatively impacts the layout and readability of the document.

![](https://i.imgur.com/l9GexvA.jpeg)

Please find a minimal example here:
<https://github.com/produnis/minimal_examples/tree/typstboxfit>

Any suggestions on how to make them fit into the box?

**Comment by mcanouil** [comment]
Which version of Quarto?

**Reply by produnis** [reply]
Sorry, quarto 1.6.37

**Comment by cderv** [comment]
Here is a reproducible example simplified
````markdown
---
title: Test
format: typst
keep-md: true
---

::: {.callout-note}

## Plots

::: {layout-ncol=5}

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})
:::

a)  Beschreiben Sie die 5 Verteilungsformen. [5 Punkte]

:::
````

![image](https://github.com/user-attachments/assets/52002384-7614-4ff2-a698-ab0cdd88cb4b)

But I really think it is a matter of having two many content for the column. Without callout would be the same

````markdown
---
title: Test
format: typst
keep-md: true
---

## Plots

::: {layout-ncol=5}

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})
:::

````
![image](https://github.com/user-attachments/assets/985c238e-0193-4616-ac71-3a51b534a1b4)

Add a 2 more and image will reduce by are not bound by the margin
```markdown
---
title: Test
format: typst
keep-md: true
---

## Plots

::: {layout-ncol=7}

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})
:::

````

![image](https://github.com/user-attachments/assets/350f2a24-483d-4e4e-a802-09ea9ab783f8)

In the callout example, there are not bound by the callout

The typst code is the follwoing 
````
#block[
#callout(
body: 
[
#block[
]
#grid(
columns: (20.0%, 20.0%, 20.0%, 20.0%, 20.0%), gutter: 1em, rows: 1,
  rect(stroke: none, width: 100%)[
#box(image("index_files\\mediabag\\svg-xml-base64,PHN2Z.svg", width: 100%))

],
  rect(stroke: none, width: 100%)[
#box(image("index_files\\mediabag\\svg-xml-base64,PHN2Z.svg", width: 100%))

],
  rect(stroke: none, width: 100%)[
#box(image("index_files\\mediabag\\svg-xml-base64,PHN2Z.svg", width: 100%))

],
  rect(stroke: none, width: 100%)[
#box(image("index_files\\mediabag\\svg-xml-base64,PHN2Z.svg", width: 100%))

],
  rect(stroke: none, width: 100%)[
#box(image("index_files\\mediabag\\svg-xml-base64,PHN2Z.svg", width: 100%))

],
)
#block[
#set enum(numbering: "a)", start: 1)
+ Beschreiben Sie die 5 Verteilungsformen. \[5 Punkte\]
]

]
, 
title: 
[
Plots
]
, 
background_color: 
rgb("#dae6fb")
, 
icon_color: 
rgb("#0758E5")
, 
icon: 
fa-info()
)
]
````

I don't know if there is a way to constraint the `#grid` to the size of the `#block[#calout()`

@gordonwoodhull do you think this is something we can improve ? If we can we probably should... 

For callout but maybe also to avoid content going in margin 🤔 We probably have still some work on [Article Layout](https://quarto.org/docs/authoring/article-layout.html) for typst

**Reply by gordonwoodhull** [reply]
It seems that using the _fraction type_ in the columns

```typst
columns: (1fr, 1fr, 1fr, 1fr, 1fr)
```
fixes this.

I suspect that when you use the ratio 

```typst
columns: (20.0%, 20.0%, 20.0%, 20.0%, 20.0%)
```

it uses portion of the page rather than using up the space of the block as we want.

**Reply by mcanouil** [reply]
It means the code in https://github.com/quarto-dev/quarto-cli/blob/2a2ddd67afc9a0d6ae211fb0d48443bf0db52bb9/src/resources/filters/layout/layout.lua would need to be fixed for Typst.

Or to convert from percentage to relative width in:

- https://github.com/quarto-dev/quarto-cli/blob/39baf4081f184fe090974295b48d1beab32cd1e4/src/resources/filters/layout/typst.lua#L51-L57

A quick&dirty hack by inserting the following at L55 fixes the issue:

```
    local all_equal = true
    for i = 2, #col_spec do
      if col_spec[i] ~= col_spec[1] then
        all_equal = false
        break
      end
    end
    if all_equal then
      for i = 1, #col_spec do
        col_spec[i] = "1fr"
      end
    end
```



| Current | Hack |
|--------|--------|
| <img width="1038" alt="image" src="https://github.com/user-attachments/assets/dbf48025-aa5d-480f-a336-97462e2a8c60"> | <img width="1038" alt="image" src="https://github.com/user-attachments/assets/0cc3adef-0960-4675-a93c-803f46163491"> | 


<details><summary>Code</summary>
<p>

````qmd
---
title: "Quarto Playground"
format: typst
keep-typ: true
---

::: {layout-ncol=7}

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})
:::


::: {.callout-note}
::: {layout-ncol=7}

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})

![]({{< placeholder format=svg >}})
:::
:::
````

</p>
</details>

**Reply by gordonwoodhull** [reply]
Converted to issue and assigned myself, but feel free to submit a PR. I’m out for a week but I’ll review when I get back.

**Reply by mcanouil** [reply]
I think I've got a simple general fix.

**Comment by gordonwoodhull** [comment, marked as answer]
Converted to issue #11578
